src/FEValues/face_integrals.jl

```diff
@@ -1,113 +1,134 @@
-    new_points = [Vec{1,T}((one(T),))] # ξ = 1
-    push!(face_quad_rule, QuadratureRule{RefLine, T}(w, new_points))
-    return FaceQuadratureRule(face_quad_rule)
+
+# Mapping from to 0D node to 1D line vertex.
+function reference_face_to_face(::Vec{N, T}, cell_T::Type{RefLine}, face::Int) where {N, T}
+    face == 1 && return Vec{1, T}(( -one(T),))
+    face == 2 && return Vec{1, T}((  one(T),))
+    throw(ArgumentError("unknown face number: $face for a cell of type $(cell_T)"))
 end
 
-function weighted_normal(::Tensor{2,1,T}, ::Type{RefLine}, face::Int) where {T}
+function weighted_normal(::Tensor{2,1,T}, cell_T::Type{RefLine}, face::Int) where {T}
     face == 1 && return Vec{1,T}((-one(T),))
     face == 2 && return Vec{1,T}(( one(T),))
-    throw(ArgumentError("unknown face number: $face"))
+    throw(ArgumentError("unknown face number: $face for a cell of type $(cell_T)"))
 end
 
 ###########################
 # All 2D RefQuadrilateral #
 ###########################
-function create_face_quad_rule(::Type{RefQuadrilateral}, w::Vector{T}, p::Vector{Vec{1, T}}) where {T}
-    n_points = length(w)
-    face_quad_rule = QuadratureRule{RefQuadrilateral, T, 2}[]
-    # Face 1
-    new_points = [Vec{2,T}((p[i][1], -one(T))) for i in 1:n_points] # ξ = t, η = -1
-    push!(face_quad_rule, QuadratureRule{RefQuadrilateral, T}(w, new_points))
-    # Face 2
-    new_points = [Vec{2,T}((one(T), p[i][1])) for i in 1:n_points] # ξ = 1, η = t
-    push!(face_quad_rule, QuadratureRule{RefQuadrilateral, T}(w, new_points))
-    # Face 3
-    new_points = [Vec{2,T}((p[i][1], one(T))) for i in 1:n_points] # ξ = t, η = 1
-    push!(face_quad_rule, QuadratureRule{RefQuadrilateral, T}(w, new_points))
-    # Face 4
-    new_points = [Vec{2,T}((-one(T), p[i][1])) for i in 1:n_points] # ξ = -1, η = t
-    push!(face_quad_rule, QuadratureRule{RefQuadrilateral, T}(w, new_points))
-    return FaceQuadratureRule(face_quad_rule)
+
+# Mapping from 1D line to 2D face of a quadrilateral.
+function reference_face_to_face(point::Vec{1, T}, cell_T::Type{RefQuadrilateral}, face::Int) where T
+    x = point[]
+    face == 1 && return Vec{2, T}(( x,          -one(T)))
+    face == 2 && return Vec{2, T}(( one(T),     x))
+    face == 3 && return Vec{2, T}(( -x,         one(T)))
+    face == 4 && return Vec{2, T}(( -one(T),    -x))
+    throw(ArgumentError("unknown face number: $face for a cell of type $(cell_T)"))
 end
 
-function weighted_normal(J::Tensor{2,2}, ::Type{RefQuadrilateral}, face::Int)
+function weighted_normal(J::Tensor{2,2}, cell_T::Type{RefQuadrilateral}, face::Int)
     @inbounds begin
         face == 1 && return Vec{2}(( J[2,1], -J[1,1]))
         face == 2 && return Vec{2}(( J[2,2], -J[1,2]))
         face == 3 && return Vec{2}((-J[2,1],  J[1,1]))
         face == 4 && return Vec{2}((-J[2,2],  J[1,2]))
     end
-    throw(ArgumentError("unknown face number: $face"))
+    throw(ArgumentError("unknown face number: $face for a cell of type $(cell_T)"))
```

Comment:
Can we do this without interpolation somehow? This kind of stuff breaks GPU compilation.

Reply:
Does throwing break GPU stuff or having a variable in the error?
I noticed unsafe stuff in #766 if that's what you mean. Also, @KnutAM suggested something like `_unsafe_weighted_normal` if it has performance benefits but I couldn't measure the benefit :smile:

Reply:
Knowing what the input face was isn't auper useful, so perhaps it can be removed. Or you can keep it as is until  this actually runs on GPU.